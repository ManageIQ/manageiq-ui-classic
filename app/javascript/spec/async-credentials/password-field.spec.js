import React from 'react';
import { mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import PasswordField from '../../components/async-credentials/password-field';

let mockChangeSpy;
let mockGetStateSpy;

jest.mock('@@ddf', () => ({
  useFormApi: () => ({
    renderForm: ([secret]) => <MockDummyComponent {...secret} />,
    change: mockChangeSpy,
    getState: mockGetStateSpy,
  }),
  componentTypes: {
    TEXT_FIELD: 'text-field',
  },
}));

const MockDummyComponent = ({
  buttonLabel,
  editMode, // eslint-disable-line
  isDisabled,
  helperText, // eslint-disable-line
  setEditMode,
  validate, // eslint-disable-line
  validateOnMount, // eslint-disable-line
  ...props
}) => <button {...props} onClick={setEditMode} disabled={isDisabled} type="button">{buttonLabel || 'Dummy'}</button>;


describe('Secret switch field component', () => {
  let initialProps;

  beforeEach(() => {
    mockChangeSpy = jest.fn();
    mockGetStateSpy = jest.fn().mockReturnValue({
      values: {},
      initialValues: { foo: 'value-foo', bar: 'value-bar', nonAsync: 'non-async' },
    });
    initialProps = {
      edit: false,
      name: 'foo',
    };
  });

  afterEach(() => {
    mockChangeSpy.mockReset();
  });

  it('should render correctly in non edit mode', () => {
    const wrapper = mount(<PasswordField {...initialProps} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly in edit mode', () => {
    const wrapper = mount(<PasswordField {...initialProps} edit />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  /**
   * Hooks are not supported in enzyme jest.
   * Instead of adding another testing utilities we can wait for while until its added or until we know enzyme will not support hooks
   * and we can use another library
   * https://github.com/airbnb/enzyme/issues/2011
   */
  it('should render correctly switch to editing', () => {
    const wrapper = mount(<PasswordField {...initialProps} edit />);
    expect(wrapper.find(MockDummyComponent)).toHaveLength(0);
    wrapper.find('button').simulate('click');
    wrapper.update();
    expect(wrapper.find(MockDummyComponent)).toHaveLength(1);
  });

  it('should render correctly reset sercret field', () => {
    const wrapper = mount(<PasswordField {...initialProps} edit />);
    wrapper.find('button').simulate('click');
    expect(wrapper.find(MockDummyComponent)).toHaveLength(1);
    wrapper.find('button').simulate('click');
    expect(mockChangeSpy).toHaveBeenCalledWith('foo', undefined);
  });
});
