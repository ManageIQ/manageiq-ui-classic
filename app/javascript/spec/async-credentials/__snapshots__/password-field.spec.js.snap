// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secret switch field component should render correctly in edit mode 1`] = `
<PasswordField
  cancelEditLabel="Cancel"
  changeEditLabel="Change"
  edit={true}
  name="foo"
>
  <FormGroup
    bsClass="form-group"
  >
    <div
      className="form-group"
    >
      <ControlLabel
        bsClass="control-label"
        srOnly={false}
      >
        <label
          className="control-label"
        />
      </ControlLabel>
      <InputGroup
        bsClass="input-group"
      >
        <span
          className="input-group"
        >
          <FormControl
            autoComplete="new-password"
            autoFocus={true}
            bsClass="form-control"
            componentClass="input"
            disabled={true}
            id="foo-password-placeholder"
            placeholder="●●●●●●●●"
            type="password"
          >
            <input
              autoComplete="new-password"
              autoFocus={true}
              className="form-control"
              disabled={true}
              id="foo-password-placeholder"
              placeholder="●●●●●●●●"
              type="password"
            />
          </FormControl>
          <InputGroupButton
            bsClass="input-group-btn"
          >
            <span
              className="input-group-btn"
            >
              <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="default"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <button
                  className="btn btn-default"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Change
                </button>
              </Button>
            </span>
          </InputGroupButton>
        </span>
      </InputGroup>
    </div>
  </FormGroup>
</PasswordField>
`;

exports[`Secret switch field component should render correctly in non edit mode 1`] = `
<PasswordField
  cancelEditLabel="Cancel"
  changeEditLabel="Change"
  edit={false}
  name="foo"
>
  <MockDummyComponent
    autoComplete="new-password"
    component="text-field"
    name="foo"
    type="password"
  >
    <button
      autoComplete="new-password"
      component="text-field"
      name="foo"
      type="button"
    >
      Dummy
    </button>
  </MockDummyComponent>
</PasswordField>
`;
